StatsModule: Optimize data selection and render

edo-ran opened the pull request:
This is performance improvement to StatsModule as part of performance improvement to My Sites.

getSiteStatsNormalizedData uses `createSelector`, which memoize the result, but because StatModule is generic and uses this selector for all types of stats the result end up never been cached because the dependent data for cache invalidation was the data of the specific stats type. This, in turn, caused the StatsModule component to render over and over even though the data didn't change.

Using the new `createSelectorPerKey` now invalidate the cache on a per-key basis which eliminates most of the wasted render calls.

There is also a change in the way `state/stats/list/reducer.js` handle received stats. Before all of the `items` subtree would have been recreated because of the use of `merge( {}, existingItems, ... )`. Now only the path that changed in site -> statType -> queryKey is recreated, the rest of the state remain exactly the same which allows `connect` HOC to detect nothing has changed and not re-render `StatsModule`.

The result is about 40ms of wasted render time instead of about 800ms of wasted render time.

## To Test
1. Visit the live branch or pull the branch
2. Transition between Reader and My Sites - use site with lots of stats

I've done the test #10326  even though we're not using `merge` any more but just to be on the safe side.

timmyc commented on client/my-sites/stats/stats-module/index.jsx:
```diff
@@ -64,6 +64,25 @@ class StatsModule extends Component {
+				}
+
+				// Special case for data prop, do deep equality check.
+				const isDeepEqual = isEqual( this.props[ key ], nextProps[ key ] );
```
Some `data` payloads include nested child objects like Clicks, Referrers, and Authors - any concerns with using `isEqual` knowing that?  Specifically Referrers can potentially be 3 levels deep:

![stats_ _the_wordpress_com_blog_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/24626914/bc9094e2-1867-11e7-87cd-52b93f2bc29c.png)

edo-ran replied:
I've measured the time it takes for the whole `shouldComponentUpdate` to run, not just the `isEqual` and got 70ms for the entire interaction of transition from Reader to My Sites which reduce the wasted render from at least 600ms down to 40ms so it still worth it.

timmyc replied:
Nice - curious was that on a site with a large amount of stats data... specifically referrers?

timmyc commented on client/lib/create-selector/index.js:
```diff
@@ -105,3 +106,28 @@ export default function createSelector( selector, getDependants = DEFAULT_GET_DE
 	}, { memoizedSelector } );
 }
 
+export function createSelectorPerKey( selector, getDependants = DEFAULT_GET_DEPENDANTS, getCacheKey = DEFAULT_GET_CACHE_KEY ) {
```
Could we possibly add some test coverage for this new function?

edo-ran replied:
You right, I've pushed commit with tests for the new method. I'm using the existing test suite of `createSelector` and added test specific for the new method.

edo-ran commented:
I have pushed a commit that fixes the way the reducer handle received stats to preserve unchanged values in `items` state tree. This allows to completely remove the `shouldComponentUpdate`.

edo-ran commented:
@aduth I've added the `createSelectorPerKey` function as we discussed in slack. I've added it as a new function instead of changing `createSelector`. I think it will be safer to change the original after I'll add statistics collection so we can see what the effect of the change is.

Do you think this is the right way or should I change the implementation of `createSelector` to work on a per key basis?

aduth commented:
A worry I have with something like this is that it encourages developers to avoid tuning their code and instead apply heavier and more granular caching. I'd be curious to measure, particularly on mobile devices, just how much we're exhausting memory versus raw CPU execution. The size of global state and memoize cache alone must be intense for limited memory devices.

To the specific example, the `getDependents` (second argument) of `getSiteStatsNormalizedData` performs an [expensive object serialization](https://github.com/Automattic/wp-calypso/blob/b0af3934d067b03914aa5e08b35cc75153d5bfab/client/state/stats/lists/utils.js#L104) and seems to me as though this might defeat the purpose of memoizing at all, because the selector itself is barely more expensive than the dependents getter which is invoked every time the selector is called.

Minor points to the implementation itself:

- Can we do a better job of reducing code duplication between the two variants?
- How can we make it less awkward that one of the two variants is the default export and the other merely a named export? Could we standardize on named exports?
- Should `createSelectorPerKey` be a separate function or defined as option to the base `createSelector` (fourth argument)?

timmyc commented:
> Can we do a better job of reducing code duplication between the two variants?

➕ on this - I noticed that as well in re-reading the code just now.

samouri commented:
I just ran into this same issue. invalidating the cache more frequently than should be happening.

I was able to get most of the way there without making a new helper function by doing something slightly counterintuitive in [this pr](https://github.com/Automattic/wp-calypso/pull/17585).  Essentially instead of making your dependent look like this: `get(state.stats.lists.items`, [ keys, go, here ]`, what if you made it look just like this: `state.stats.lists.items`.

alisterscott commented:
Closing as this PR is stale - please reopen if necessary
